File: C++/2007. Find Original Array From Doubled Array.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
 public:
  vector<int> findOriginalArray(vector<int>& changed) {
    sort(changed.begin(), changed.end());
    int n = changed.size();
    vector<int> res;
    if (n % 2 == 1) {
      return {};
    }

    map<int, int> mp;
    for (int i = 0; i < n; i++) {
      if (changed[i] % 2 == 1) {
        if (mp.find(changed[i]) != mp.end()) {
          mp[changed[i]]++;
        } else {
          mp[changed[i]] = 1;
        }
      } else {
        if (mp.find(changed[i] / 2) != mp.end()) {
          if (mp[changed[i] / 2] == 1) {
            mp.erase(changed[i] / 2);
          } else {
            mp[changed[i] / 2]--;
          }
          res.push_back(changed[i] / 2);
        } else {
          if (mp.find(changed[i]) != mp.end()) {
            mp[changed[i]]++;
          } else {
            mp[changed[i]] = 1;
          }
        }
      }
    }
    if (res.size() == n / 2) {
      return res;
    } else {
      return {};
    }
  }
};

File: JavaScript/2007. Find Original Array From Doubled Array.js
/**
 * @param {number[]} changed
 * @return {number[]}
 */
var findOriginalArray = function (changed) {
  let n = changed.length;
  if (n % 2 == 1) {
    return [];
  }

  changed.sort((a, b) => a - b);

  let can = true;
  const res = [];
  const mp = new Map();
  let pair = 0;
  for (let i = 0; i < n; i++) {
    if (changed[i] % 2 == 1) {
      if (mp.has(changed[i])) {
        mp.set(changed[i], mp.get(changed[i]) + 1);
      } else {
        mp.set(changed[i], 1);
      }
    } else {
      if (mp.has(Math.floor(changed[i] / 2))) {
        if (mp.get(Math.floor(changed[i] / 2)) === 1) {
          mp.delete(Math.floor(changed[i] / 2));
        } else {
          mp.set(
            Math.floor(changed[i] / 2),
            mp.get(Math.floor(changed[i] / 2)) - 1
          );
        }
        res.push(Math.floor(changed[i] / 2));
      } else {
        if (mp.has(changed[i])) {
          mp.set(changed[i], mp.get(changed[i]) + 1);
        } else {
          mp.set(changed[i], 1);
        }
      }
    }
  }
  if (res.length === n / 2) {
    return res;
  } else {
    return [];
  }
};
